packages/eip-7702-internal-rpc-middleware/src/utils.ts

```diff
@@ -0,0 +1,85 @@
+import type { Hex } from '@metamask/utils';
+import { isHexAddress } from '@metamask/utils';
+
+/**
+ * Validates address format, checks user eth_accounts permissions, and normalizes to lowercase.
+ *
+ * @param address - The Ethereum address to validate and normalize.
+ * @param origin - The origin string for permission checking.
+ * @param getPermittedAccountsForOrigin - Function to retrieve permitted accounts for the origin.
+ * @returns A normalized (lowercase) hex address if valid and authorized.
+ * @throws JsonRpcError with unauthorized error if the requester doesn't have permission to access the address.
+ * @throws JsonRpcError with invalid params if the address format is invalid.
+ */
+export async function validateAndNormalizeAddress(
+  address: Hex,
+  origin: string,
+  getPermittedAccountsForOrigin: (origin: string) => Promise<string[]>,
+): Promise<Hex> {
+  if (
+    typeof address === 'string' &&
+    address.length > 0 &&
+    isHexAddress(address)
+  ) {
+    // Ensure that an "unauthorized" error is thrown if the requester
+    // does not have the `eth_accounts` permission.
+    const accounts = await getPermittedAccountsForOrigin(origin);
+
+    // Validate and convert each account address to normalized Hex
+    const normalizedAccounts: string[] = accounts.map((accountAddress) =>
+      accountAddress.toLowerCase(),
+    );
+
+    const normalizedAddress = address.toLowerCase();
+
+    if (normalizedAccounts.includes(normalizedAddress)) {
+      // we know that normalizedAddress is a valid Hex string because of isHexAddress
+      return normalizedAddress as Hex;
+    }
+
+    throw providerErrors.unauthorized();
+  }
+
+  throw rpcErrors.invalidParams({
+    message: `Invalid parameters: must provide an Ethereum address.`,
+  });
+}
+
+/**
+ * Validates parameters against a Superstruct schema and throws an error if validation fails.
+ *
+ * @param value - The value to validate against the struct schema.
+ * @param struct - The Superstruct schema to validate against.
+ * @throws JsonRpcError with invalid params if the value doesn't match the struct schema.
+ */
+export function validateParams<ParamsType>(
+  value: unknown | ParamsType,
+  struct: Struct<ParamsType>,
+): asserts value is ParamsType {
+  const [error] = validate(value, struct);
+
+  if (error) {
+    throw rpcErrors.invalidParams(
+      formatValidationError(error, 'Invalid parameters'),
+    );
+  }
+}
+
+/**
+ * Formats a Superstruct validation error into a human-readable string.
+ *
+ * @param error - The Superstruct validation error to format.
+ * @param message - The base error message to prepend to the formatted details.
+ * @returns A formatted error message string with validation failure details.
+ */
+function formatValidationError(error: StructError, message: string): string {
+  return `${message}\n\n${error
+    .failures()
+    .map(
+      (f) => `${f.path.join(' > ')}${f.path.length ? ' - ' : ''}${f.message}`,
+    )
+    .join('\n')}`;
+}
```

Comment:
I feel like I've seen this pattern in a few other parts of the codebase already,  or code shaped like this at least

Reply:
There is the same function in the `eip-5792-middleware` package this is copied from there. Found similar thing in `user-operation-controller` as well. But its not in any package/lib that is shared across packages. 

Should I extract this function to some kind of common utils? No sure to which package this would be most appropriate.